mentat/embeddings.py

```diff
@@ -0,0 +1,138 @@
+import asyncio
+import gzip
+import json
+import os
+from pathlib import Path
+
+import numpy as np
+
+from .code_file import CodeFile
+from .config_manager import mentat_dir_path
+from .llm_api import call_embedding_api, count_tokens
+from .session_stream import SESSION_STREAM
+from .utils import sha256
+
+EMBEDDING_MODEL = "text-embedding-ada-002"
+EMBEDDING_MAX_TOKENS = 8192
+
+
+class EmbeddingsDatabase:
+    # { sha256 : [ 1536 floats ] }
+    _dict: dict[str, list[float]] = dict[str, list[float]]()
+
+    def __init__(self, output_dir: Path | None = None):
+        if output_dir is None:
+            output_dir = mentat_dir_path
+        os.makedirs(output_dir, exist_ok=True)
+        self.path = Path(output_dir) / "embeddings.json.gz"
+        if self.path.exists():
+            with gzip.open(self.path, "rt") as f:
+                self._dict = json.load(f)
+
+    def save(self):
+        with gzip.open(self.path, "wt") as f:
+            json.dump(self._dict, f)
+
+    def __getitem__(self, key: str) -> list[float]:
+        return self._dict[key]
+
+    def __setitem__(self, key: str, value: list[float]):
+        self._dict[key] = value
+
+    def __contains__(self, key: str) -> bool:
+        return key in self._dict
+
+
+database = EmbeddingsDatabase()
+
+
+def _batch_ffd(data: dict[str, int], batch_size: int) -> list[list[str]]:
+    """Batch files using the First Fit Decreasing algorithm."""
+    # Sort the data by the length of the strings in descending order
+    sorted_data = sorted(data.items(), key=lambda x: x[1], reverse=True)
+    batches = list[list[str]]()
+    for key, value in sorted_data:
+        # Place each item in the first batch that it fits in
+        placed = False
+        for batch in batches:
+            if sum(data[k] for k in batch) + value <= batch_size:
+                batch.append(key)
+                placed = True
+                break
+        if not placed:
+            batches.append([key])
+    return batches
+
+
+def _cosine_similarity(v1: list[float], v2: list[float]) -> float:
+    """Calculate the cosine similarity between two vectors."""
+    dot_product = np.dot(v1, v2)
+    norm_v1 = np.linalg.norm(v1)
+    norm_v2 = np.linalg.norm(v2)
+    return dot_product / (norm_v1 * norm_v2)
+
+
+async def _count_feature_tokens(features: list[CodeFile], model: str) -> list[int]:
```

Comment:
Is this not a duplicate of the other count feature tokens function? And just like my other comment, do we need a semaphore? If not, both of these functions wouldn't be needed

Reply:
Yes - just combined these into a single function `count_feature_tokens` in CodeFile.py (for import reasons).